### PR TITLE
[serve] Update API stability annotations for 2.7 release

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Updates/cleans up API stability annotations in anticipation of the 2.7 release. Changes aside from upgrading stabilities:

- Deprecate `serve.build` and `BuiltApplication`.
- Move `DeploymentConfig` and `ReplicaConfig` to a private file without an annotation (they're not publicly-exposed APIs).
- `s/get_global_client/_get_global_client` (also not a publicly-exposed API).
- Remove annotation from `RequestContext` and make it `_RequestContext` (also not a publicly-exposed API).
- Add `ProxyLocation` type to supersede `DeploymentMode`.
- Add docstrings to `HTTPOptions` and `ProxyLocation`.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by edoakes** [comment]
@zcin PTAL

File: python/ray/serve/context.py [hunk 1]
```diff
@@ -22,20 +22,27 @@
...
-@PublicAPI(stability="alpha")
+@DeveloperAPI
 @dataclass
 class ReplicaContext:
-    """Stores data for Serve API calls from within deployments."""
+    """Stores data for Serve API calls from within deployments.
+
+    Fields:
 
+        - app_name: name of the application this is running in.
```
Comment (shrekris-anyscale) [inline python/ray/serve/context.py:32]: > ... this is running in.

This wording is a bit confusing. Could you reword it to clarify what "this" is?

File: python/ray/serve/schema.py [hunk 2]
```diff
@@ -597,24 +596,24 @@ class ServeDeploySchema(BaseModel):
...
-        default=DeploymentMode.EveryNode,
+    proxy_location: ProxyLocation = Field(
+        default=ProxyLocation.EveryNode,
         description=(
-            "The location of HTTP servers.\n"
-            '- "EveryNode" (default): start one HTTP server per node.\n'
-            '- "HeadOnly": start one HTTP server on the head node.\n'
-            '- "NoServer": disable HTTP server.'
+            "Config for where to run proxies for ingress traffic to the cluster.\n"
+            '- "Disabled": disable the proxies entirely.'
```
Comment (shrekris-anyscale) [inline python/ray/serve/schema.py:603]:
```suggestion
            '- "Disabled": disable the proxies entirely.\n'
```

File: python/ray/serve/schema.py [hunk 3]
```diff
@@ -864,12 +863,12 @@ class ServeInstanceDetails(BaseModel, extra=Extra.forbid):
...
     )
-    proxy_location: Optional[DeploymentMode] = Field(
+    proxy_location: Optional[ProxyLocation] = Field(
         description=(
-            "The location of HTTP servers.\n"
-            '- "EveryNode": start one HTTP server per node.\n'
-            '- "HeadOnly": start one HTTP server on the head node.\n'
-            '- "NoServer": disable HTTP server.'
+            "Config for where to run proxies for ingress traffic to the cluster.\n"
+            '- "Disabled": disable the proxies entirely.'
```
Comment (shrekris-anyscale) [inline python/ray/serve/schema.py:869]:
```suggestion
            '- "Disabled": disable the proxies entirely.\n'
```

File: python/ray/serve/_private/config.py [hunk 4]
```diff
@@ -0,0 +1,634 @@
+import inspect
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/config.py:1]: nope pure move

File: python/ray/serve/context.py (hunk 1 above)
Comment (edoakes) [reply python/ray/serve/context.py:32]: improved

File: python/ray/serve/schema.py [hunk 5]
```diff
@@ -538,13 +537,13 @@ class gRPCOptionsSchema(BaseModel):
     )
```
Comment (GeneDer) [inline python/ray/serve/schema.py:537]: If `ServeDeploySchema` is marked as stable, should we also mark `gRPCOptionsSchema` as stable as well? 🤔

**Review by GeneDer — APPROVED** [review]
LGTM!

File: python/ray/serve/schema.py (hunk 5 above)
Comment (edoakes) [reply python/ray/serve/schema.py:537]: No it's fine to have sub-fields that can change if they're marked.

gRPC is still experimental, it should be marked alpha

**Review by shrekris-anyscale — APPROVED** [review]
Nice work!

File: doc/source/serve/key-concepts.md [hunk 6]
```diff
@@ -75,7 +75,7 @@ serve.run(driver)
 ## Ingress Deployment (HTTP handling)
 
 A Serve application can consist of multiple deployments that can be combined to perform model composition or complex business logic.
-However, there is always one "top-level" deployment, the one that will be passed to `serve.run` or `serve.build` to deploy the application.
+However, there is always one "top-level" deployment, the one that will be passed to `serve.run` to deploy the application.
```
Comment (angelinalg) [inline doc/source/serve/key-concepts.md:78]:
```suggestion
However, one deployment is always the "top-level" one that is passed to `serve.run` to deploy the application.
```

File: doc/source/serve/key-concepts.md [hunk 7]
```diff
@@ -75,7 +75,7 @@ serve.run(driver)
...
 This deployment is called the "ingress deployment" because it serves as the entrypoint for all traffic to the application.
 Often, it will then route to other deployments or call into them using the `ServeHandle` API and compose the results before returning to the user.
```
Comment (angelinalg) [inline doc/source/serve/key-concepts.md:80]:
```suggestion
Often, it then routes to other deployments or calls into them using the `ServeHandle` API, and composes the results before returning to the user.
```

File: python/ray/serve/_private/config.py [hunk 8]
```diff
@@ -0,0 +1,634 @@
...
+    else:
+        return False
+
+
+class DeploymentConfig(BaseModel):
+    """Internal datastructure wrapping config options for a deployment.
+
+    Args:
+        num_replicas (Optional[int]): The number of processes to start up that
+            will handle requests to this deployment. Defaults to 1.
```
Comment (angelinalg) [inline python/ray/serve/_private/config.py:57]:
```suggestion
            handles requests to this deployment. Defaults to 1.
```

File: python/ray/serve/_private/config.py [hunk 9]
```diff
@@ -0,0 +1,634 @@
...
+        max_concurrent_queries (Optional[int]): The maximum number of queries
+            that will be sent to a replica of this deployment without receiving
```
Comment (angelinalg) [inline python/ray/serve/_private/config.py:59]:
```suggestion
            that is sent to a replica of this deployment without receiving
```

File: python/ray/serve/_private/config.py [hunk 10]
```diff
@@ -0,0 +1,634 @@
...
+            a response. Defaults to 100.
+        user_config (Optional[Any]): Arguments to pass to the reconfigure
+            method of the deployment. The reconfigure method is called if
+            user_config is not None. Must be json-serializable.
```
Comment (angelinalg) [inline python/ray/serve/_private/config.py:63]:
```suggestion
            user_config is not None. Must be JSON-serializable.
```

File: python/ray/serve/_private/config.py [hunk 11]
```diff
@@ -0,0 +1,634 @@
...
+        graceful_shutdown_wait_loop_s (Optional[float]): Duration
+            that deployment replicas will wait until there is no more work to
```
Comment (angelinalg) [inline python/ray/serve/_private/config.py:65]:
```suggestion
            that deployment replicas wait until there is no more work to
```

File: python/ray/serve/_private/config.py [hunk 12]
```diff
@@ -0,0 +1,634 @@
...
+            be done before shutting down.
+        graceful_shutdown_timeout_s (Optional[float]):
+            Controller waits for this duration to forcefully kill the replica
+            for shutdown.
+        health_check_period_s (Optional[float]):
+            Frequency at which the controller will health check replicas.
```
Comment (angelinalg) [inline python/ray/serve/_private/config.py:71]:
```suggestion
            Frequency at which the controller health checks replicas.
```

File: python/ray/serve/_private/config.py [hunk 13]
```diff
@@ -0,0 +1,634 @@
...
+        health_check_timeout_s (Optional[float]):
+            Timeout that the controller will wait for a response from the
```
Comment (angelinalg) [inline python/ray/serve/_private/config.py:73]:
```suggestion
            Timeout that the controller waits for a response from the
```

File: python/ray/serve/_private/config.py [hunk 14]
```diff
@@ -0,0 +1,634 @@
...
+    Provides five main properties (see property docstrings for more info):
+        deployment_def: the code, or a reference to the code, that this
+            replica should run.
+        init_args: the deployment_def's init_args.
+        init_kwargs: the deployment_def's init_kwargs.
+        ray_actor_options: the Ray actor options to pass into the replica's
+            actor.
+        resource_dict: contains info on this replica's actor's resource needs.
+
+    Offers a serialized equivalent (e.g. serialized_deployment_def) for
```
Comment (angelinalg) [inline python/ray/serve/_private/config.py:263]:
```suggestion
    Offers a serialized equivalent (e.g., serialized_deployment_def) for
```

File: python/ray/serve/_private/config.py [hunk 15]
```diff
@@ -0,0 +1,634 @@
...
+    deployment_def, init_args, and init_kwargs. Deserializes these properties
+    when they're first accessed, if they were not passed in directly through
+    create().
+
+    Use the classmethod create() to make a ReplicaConfig with the deserialized
+    properties.
+
+    Note: overwriting or setting any property after the ReplicaConfig has been
+    constructed is currently undefined behavior. The config's fields should not
```
Comment (angelinalg) [inline python/ray/serve/_private/config.py:272]:
```suggestion
    constructed is currently undefined behavior. The config's fields shouldn't
```

File: python/ray/serve/_private/config.py [hunk 16]
```diff
@@ -0,0 +1,634 @@
...
+                    )
+
+                # Validate that the replica actor fits in the first bundle.
+                if i == 0:
+                    bundle_cpu = bundle.get("CPU", 0)
+                    replica_actor_num_cpus = self.ray_actor_options.get("num_cpus", 0)
+                    if bundle_cpu < replica_actor_num_cpus:
+                        raise ValueError(
+                            "When using `placement_group_bundles`, the replica actor "
+                            "will be placed in the first bundle, so the resource "
```
Comment (angelinalg) [inline python/ray/serve/_private/config.py:510]:
```suggestion
                            "is placed in the first bundle, so the resource "
```

File: python/ray/serve/_private/config.py [hunk 17]
```diff
@@ -0,0 +1,634 @@
...
+                            f"{replica_actor_num_cpus} but the bundle only has "
+                            f"{bundle_cpu} `CPU` specified."
+                        )
+
+                    bundle_gpu = bundle.get("GPU", 0)
+                    replica_actor_num_gpus = self.ray_actor_options.get("num_gpus", 0)
+                    if bundle_gpu < replica_actor_num_gpus:
+                        raise ValueError(
+                            "When using `placement_group_bundles`, the replica actor "
+                            "will be placed in the first bundle, so the resource "
```
Comment (angelinalg) [inline python/ray/serve/_private/config.py:522]:
```suggestion
                            "is placed in the first bundle, so the resource "
```

File: python/ray/serve/_private/config.py [hunk 18]
```diff
@@ -0,0 +1,634 @@
...
+
+                    replica_actor_resources = self.ray_actor_options.get(
+                        "resources", {}
+                    )
+                    for actor_resource, actor_value in replica_actor_resources.items():
+                        bundle_value = bundle.get(actor_resource, 0)
+                        if bundle_value < actor_value:
+                            raise ValueError(
+                                "When using `placement_group_bundles`, the replica "
+                                "actor will be placed in the first bundle, so the "
```
Comment (angelinalg) [inline python/ray/serve/_private/config.py:537]:
```suggestion
                                "actor is placed in the first bundle, so the "
```

File: python/ray/serve/api.py [hunk 19]
```diff
@@ -90,47 +90,15 @@ def start(
           script. If set, the instance will live on the Ray cluster until it is
```
Comment (angelinalg) [inline python/ray/serve/api.py:90]:
```suggestion
          script. If set, the instance lives on the Ray cluster until it is
```

File: python/ray/serve/config.py [hunk 20]
```diff
@@ -120,618 +98,74 @@ def get_downscale_smoothing_factor(self) -> PositiveFloat:
...
+            raise ValueError(f"Unrecognized `ProxyLocation`: {v}.")
 
 
-@PublicAPI(stability="beta")
+@PublicAPI(stability="stable")
 class HTTPOptions(pydantic.BaseModel):
-    # Documentation inside serve.start for user's convenience.
+    """HTTP options for the proxies. Supported fields:
+
+    - host: Host that the proxies will listen for HTTP on. Defaults to
```
Comment (angelinalg) [inline python/ray/serve/config.py:145]:
```suggestion
    - host: Host that the proxies listens for HTTP on. Defaults to
```

File: python/ray/serve/config.py [hunk 21]
```diff
@@ -120,618 +98,74 @@ def get_downscale_smoothing_factor(self) -> PositiveFloat:
...
+      "127.0.0.1". To expose Serve publicly, you probably want to set
+      this to "0.0.0.0".
+    - port: Port that the proxies will listen for HTTP on. Defaults to 8000.
```
Comment (angelinalg) [inline python/ray/serve/config.py:148]:
```suggestion
    - port: Port that the proxies listen for HTTP on. Defaults to 8000.
```

File: python/ray/serve/config.py [hunk 22]
```diff
@@ -120,618 +98,74 @@ def get_downscale_smoothing_factor(self) -> PositiveFloat:
...
+    - root_path: An optional root path to mount the serve application
+      (for example, "/prefix"). All deployment routes will be prefixed
```
Comment (angelinalg) [inline python/ray/serve/config.py:150]:
```suggestion
      (for example, "/prefix"). All deployment routes are prefixed
```

**Review by angelinalg — APPROVED** [review]
I dabbled in some Python files. Let me know if this is more of a nuisance, than helpful.

**Review by sihanwang41 — APPROVED** [review]

File: doc/source/serve/api/index.md [hunk 23]
```diff
@@ -44,31 +44,44 @@ This is fixed by added custom filename mappings in `source/conf.py` (look for "a
...
    :nosignatures:
    :toctree: doc/
 
-   serve.run
-   serve.delete
    serve.start
-   serve.shutdown
-   serve.status
+   serve.config.ProxyLocation
+   serve.config.HTTPOptions
```
Comment (zcin) [inline doc/source/serve/api/index.md:70]: Nice! Could we add `serve.config.gRPCOptions` as well?

File: doc/source/serve/api/index.md [hunk 24]
```diff
@@ -44,31 +44,44 @@ This is fixed by added custom filename mappings in `source/conf.py` (look for "a
...
+```{eval-rst}
+.. autosummary::
+   :nosignatures:
+   :toctree: doc/
+
+   serve.get_replica_context
+   serve.context.ReplicaContext
+   serve.get_multiplexed_model_id
+   serve.get_app_handle
+   serve.get_deployment_handle
```
Comment (zcin) [inline doc/source/serve/api/index.md:84]: Will you be adding the new `DeploymentHandle` to api refs in a separate PR?
image

File: python/ray/serve/config.py [hunk 25]
```diff
@@ -120,618 +98,74 @@ def get_downscale_smoothing_factor(self) -> PositiveFloat:
...
-    def _validate_placement_group_options(self) -> None:
-        if (
-            self.placement_group_strategy is not None
-            and self.placement_group_strategy not in VALID_PLACEMENT_GROUP_STRATEGIES
-        ):
-            raise ValueError(
-                f"Invalid placement group strategy '{self.placement_group_strategy}'. "
-                f"Supported strategies are: {VALID_PLACEMENT_GROUP_STRATEGIES}."
-            )
+class ProxyLocation(str, Enum):
```
Comment (zcin) [inline python/ray/serve/config.py:111]: Why is this marked as DeveloperAPI, if it's a public parameter in `serve.start`? (is there a plan to promote this to PublicAPI in the future?)

File: python/ray/serve/schema.py [hunk 26]
```diff
@@ -597,24 +596,24 @@ class ServeDeploySchema(BaseModel):
...
-    proxy_location: DeploymentMode = Field(
-        default=DeploymentMode.EveryNode,
+    proxy_location: ProxyLocation = Field(
+        default=ProxyLocation.EveryNode,
         description=(
-            "The location of HTTP servers.\n"
-            '- "EveryNode" (default): start one HTTP server per node.\n'
-            '- "HeadOnly": start one HTTP server on the head node.\n'
-            '- "NoServer": disable HTTP server.'
+            "Config for where to run proxies for ingress traffic to the cluster.\n"
```
Comment (zcin) [inline python/ray/serve/schema.py:602]:
```suggestion
            "Config for where to run proxies to receive ingress traffic to the cluster.\n"
```

Comment (zcin) [reply python/ray/serve/schema.py:602]: Also I think the formatting is a little borked... maybe try double newlines at the end?
Screen Shot 2023-08-31 at 3 22 07 PM

File: doc/source/serve/api/index.md (hunk 23 above)
Comment (zcin) [reply doc/source/serve/api/index.md:70]: On a similar note, could we add `schema.gRPCOptionsSchema` in the config schema section?
Screen Shot 2023-08-31 at 3 25 19 PM

Comment (edoakes) [reply doc/source/serve/api/index.md:70]: leaving these to @GeneDer

File: doc/source/serve/api/index.md (hunk 24 above)
Comment (edoakes) [reply doc/source/serve/api/index.md:84]: yep, will add them separately

File: python/ray/serve/config.py (hunk 25 above)
Comment (edoakes) [reply python/ray/serve/config.py:111]: fair point, let me just upgrade it

File: python/ray/serve/schema.py (hunk 26 above)
Comment (edoakes) [reply python/ray/serve/schema.py]: documenting the options is fully redundant because it links to the `ProxyLocation` ref, so just removing it here

**Review by zcin — APPROVED** [review]
